File: src/python/_2265_Count_Nodes_Equal_to_Average_of_Subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        def dfs(root):
            total = root.val
            count = 1

            if root.left:
                left = dfs(root.left)
                total += left[0]
                count += left[1]

            if root.right:
                right = dfs(root.right)
                total += right[0]
                count += right[1]

            nonlocal ans

            if total // count == root.val:
                ans += 1

            return (total, count)

        ans = 0

        dfs(root)

        return ans

File: src/java/_2265_Count_Nodes_Equal_to_Average_of_Subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int ans = 0;

    private int[] dfs(TreeNode root) {
        int total = root.val, count = 1;

        if (root.left != null) {
            int[] left = dfs(root.left);
            total += left[0];
            count += left[1];
        }

        if (root.right != null) {
            int[] right = dfs(root.right);
            total += right[0];
            count += right[1];
        }

        if (total / count == root.val)
            this.ans += 1;

        return new int[] { total, count };
    }

    public int averageOfSubtree(TreeNode root) {
        dfs(root);
        return this.ans;
    }
}
